File: 2023/day4/Scratchcards.py
import sys
import re

def calc_matches(card):
    match = re.match(r"Card\s+(\d+):(.+)\|(.+)", card)
    if match:
        card_num, winning_numbers, my_numbers = match.groups()
        winning_numbers_set = set(map(int, winning_numbers.split()))
        my_numbers_set = set(map(int, my_numbers.split()))
        matching_numbers = winning_numbers_set.intersection(my_numbers_set)
        return len(matching_numbers)
    return 0

def part1(cards):
    total_points = 0
    for card in cards:
        matches = calc_matches(card)
        if matches:
            total_points += 2 ** (calc_matches(card) - 1)
    print(f"Part 1: {total_points}")

def part2(cards):
    total_cards = 0
    cards_dict = {}
    for card in cards:
        card_match = re.match(r"Card\s+(\d+):", card)
        if card_match:
            card_num = card_match.groups()[0]
            card_num = int(card_num)
            cards_dict[card_num] = cards_dict.get(card_num, 0) + 1
            matches = calc_matches(card)
            for i in range(1, matches + 1):
                cards_dict[card_num + i] = cards_dict.get(card_num + i, 0) + cards_dict[card_num]

    total_sum = sum(cards_dict.values())
    print(f"Part 2: {total_sum}")

def main():
    with open(sys.argv[1], 'r') as file:
        cards = file.read().strip().split('\n')

    part1(cards)
    part2(cards)

if __name__ == "__main__":
    main()

File: 2023/day4/Scratchcards.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func calcMatches(card string) int {
	re := regexp.MustCompile(`Card\s+(\d+):(.+)\|(.+)`)
	match := re.FindStringSubmatch(card)
	if len(match) > 0 {
		winningNumbers, myNumbers := match[2], match[3]
		winningNumbersSet := make(map[int]struct{})
		myNumbersSet := make(map[int]struct{})

		for _, numStr := range strings.Fields(winningNumbers) {
			num, _ := strconv.Atoi(numStr)
			winningNumbersSet[num] = struct{}{}
		}

		for _, numStr := range strings.Fields(myNumbers) {
			num, _ := strconv.Atoi(numStr)
			myNumbersSet[num] = struct{}{}
		}

		matchingNumbers := make(map[int]struct{})
		for num := range winningNumbersSet {
			if _, ok := myNumbersSet[num]; ok {
				matchingNumbers[num] = struct{}{}
			}
		}

		return len(matchingNumbers)
	}
	return 0
}

func part1(cards []string) {
	totalPoints := 0
	for _, card := range cards {
		matches := calcMatches(card)
		if matches > 0 {
			totalPoints += 1 << (matches - 1)
		}
	}
	fmt.Printf("Part 1: %d\n", totalPoints)
}

func part2(cards []string) {
	cardsDict := make(map[int]int)
	for _, card := range cards {
		cardNumMatch := regexp.MustCompile(`Card\s+(\d+):`).FindStringSubmatch(card)
		if len(cardNumMatch) > 0 {
			cardNum, _ := strconv.Atoi(cardNumMatch[1])
			cardsDict[cardNum] += 1
			matches := calcMatches(card)
			for i := 1; i <= matches; i++ {
				cardsDict[cardNum+i] += cardsDict[cardNum]
			}
		}
	}

	totalSum := 0
	for _, count := range cardsDict {
		totalSum += count
	}

	fmt.Printf("Part 2: %d\n", totalSum)
}

func main() {
	file, err := os.Open(os.Args[1])
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	var cards []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		cards = append(cards, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	part1(cards)
	part2(cards)
}
